File: apps/flutter_parent/lib/utils/common_widets/user_avatar.dart
/// Copyright (C) 2019 - present Instructure, Inc.
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, version 3 of the License.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import 'package:flutter/material.dart';
import 'package:flutter_parent/models/user.dart';

class UserAvatar extends StatelessWidget {
  final User _user;

  UserAvatar(this._user, {Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    if (_user.avatarUrl != null) {
      return Container(
        margin: EdgeInsets.only(top: 16.0),
        decoration: BoxDecoration(
            shape: BoxShape.circle,
            image: DecorationImage(image: NetworkImage(_user.avatarUrl), fit: BoxFit.contain)),
        width: 48,
        height: 48,
        alignment: AlignmentDirectional.centerStart,
      );
    } else {
      return CircleAvatar(child: Text(getUserInitials(_user)));
    }
  }

  static String getUserInitials(User _user) {
    if (_user.shortName == null || _user.shortName.isEmpty) return '?';

    var name = _user.shortName;

    // Take the first letter of each word, uppercase it, and put it into a list
    var initials = name.trim().split(RegExp(r"\s+")).map((it) => it.toUpperCase()[0]).toList();

    if (initials.length == 2) {
      // We have two initials, put them together into one string and return it
      return initials.join('');
    } else {
      // Just take the first initial if we don't have exactly two
      return initials[0];
    }
  }

}

File: libs/pandautils/src/main/java/com/instructure/pandautils/compose/composables/UserAvatar.kt
/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.pandautils.compose.composables

import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.core.graphics.drawable.toBitmap
import com.bumptech.glide.integration.compose.ExperimentalGlideComposeApi
import com.bumptech.glide.integration.compose.GlideImage
import com.instructure.pandautils.R
import com.instructure.pandautils.utils.ProfileUtils


@OptIn(ExperimentalGlideComposeApi::class)
@Composable
fun UserAvatar(
    imageUrl: String?,
    name: String,
    anonymous: Boolean = false,
    modifier: Modifier = Modifier
) {
    val model: Any? = when {
        anonymous -> R.drawable.ic_user_avatar
        ProfileUtils.shouldLoadAltAvatarImage(imageUrl) -> ProfileUtils.createAvatarDrawable(
            context = LocalContext.current,
            userName = name,
            borderWidth = with(LocalDensity.current) {
                dimensionResource(id = R.dimen.avatar_border_width_thin).toPx().toInt()
            }
        ).toBitmap()
        else -> imageUrl
    }

    GlideImage(
        model = model,
        contentDescription = null,
        modifier = modifier.clip(CircleShape),
        contentScale = ContentScale.Crop
    ) {
        it.placeholder(R.drawable.recipient_avatar_placeholder)
    }
}

@Preview(showBackground = true)
@Composable
fun UserAvatarPreview() {
    UserAvatar(
        imageUrl = null,
        name = "Test User",
        modifier = Modifier.size(100.dp)
    )
}
